// @pipcook/pipcook-cli/dist/utils/request.js
"use strict";
var __createBinding = (this && this.__createBinding) || (Object.create ? (function(o, m, k, k2) {
    if (k2 === undefined) k2 = k;
    Object.defineProperty(o, k2, { enumerable: true, get: function() { return m[k]; } });
}) : (function(o, m, k, k2) {
    if (k2 === undefined) k2 = k;
    o[k2] = m[k];
}));
var __setModuleDefault = (this && this.__setModuleDefault) || (Object.create ? (function(o, v) {
    Object.defineProperty(o, "default", { enumerable: true, value: v });
}) : function(o, v) {
    o["default"] = v;
});
var __importStar = (this && this.__importStar) || function (mod) {
    if (mod && mod.__esModule) return mod;
    var result = {};
    if (mod != null) for (var k in mod) if (k !== "default" && Object.hasOwnProperty.call(mod, k)) __createBinding(result, mod, k);
    __setModuleDefault(result, mod);
    return result;
};
var __awaiter = (this && this.__awaiter) || function (thisArg, _arguments, P, generator) {
    function adopt(value) { return value instanceof P ? value : new P(function (resolve) { resolve(value); }); }
    return new (P || (P = Promise))(function (resolve, reject) {
        function fulfilled(value) { try { step(generator.next(value)); } catch (e) { reject(e); } }
        function rejected(value) { try { step(generator["throw"](value)); } catch (e) { reject(e); } }
        function step(result) { result.done ? resolve(result.value) : adopt(result.value).then(fulfilled, rejected); }
        step((generator = generator.apply(thisArg, _arguments || [])).next());
    });
};
var __importDefault = (this && this.__importDefault) || function (mod) {
    return (mod && mod.__esModule) ? mod : { "default": mod };
};
Object.defineProperty(exports, "__esModule", { value: true });
exports.listen = exports.uploadFile = exports.getFile = exports.get = exports.del = exports.put = exports.post = void 0;
const qs = __importStar(require("querystring"));
const util_1 = require("util");
const axios_1 = __importDefault(require("axios"));
const fs_extra_1 = __importDefault(require("fs-extra"));
const eventsource_1 = __importDefault(require("eventsource"));
const common_1 = require("./common");
const form_data_1 = __importDefault(require("form-data"));
function createGeneralRequest(agent) {
    return (...args) => __awaiter(this, void 0, void 0, function* () {
        var _a, _b;
        try {
            const response = yield agent(...args);
            if (response.data.status === true) {
                return response.data.data;
            }
        }
        catch (err) {
            if ((_b = (_a = err === null || err === void 0 ? void 0 : err.response) === null || _a === void 0 ? void 0 : _a.data) === null || _b === void 0 ? void 0 : _b.message) {
                common_1.logger.fail(err.response.data.message, false);
            }
            else {
                console.error('daemon is not started, run "pipcook daemon start"');
            }
            return process.exit(1);
        }
    });
}
exports.post = (host, body, params) => __awaiter(void 0, void 0, void 0, function* () { return createGeneralRequest(axios_1.default.post)(host, body, params); });
exports.put = (host, body, params) => __awaiter(void 0, void 0, void 0, function* () { return createGeneralRequest(axios_1.default.put)(host, body, params); });
exports.del = (host) => __awaiter(void 0, void 0, void 0, function* () { return createGeneralRequest(axios_1.default.delete)(host); });
exports.get = (host, params) => __awaiter(void 0, void 0, void 0, function* () {
    const uri = `${host}?${qs.stringify(params)}`;
    return createGeneralRequest(axios_1.default.get)(uri);
});
exports.getFile = (host, params) => __awaiter(void 0, void 0, void 0, function* () {
    const resp = yield axios_1.default({
        method: 'GET',
        url: `${host}?${qs.stringify(params)}`,
        responseType: 'stream'
    });
    return resp.data;
});
exports.uploadFile = (host, file, params) => __awaiter(void 0, void 0, void 0, function* () {
    const stream = fs_extra_1.default.createReadStream(file);
    const form = new form_data_1.default();
    for (const key in params) {
        if (params[key]) {
            form.append(key, params[key]);
        }
    }
    form.append('file', stream);
    const getLength = util_1.promisify(form.getLength.bind(form));
    const length = yield getLength();
    const headers = Object.assign({ 'Content-Length': length }, form.getHeaders());
    return axios_1.default.post(host, form, { headers });
});
exports.listen = (host, params, handlers) => __awaiter(void 0, void 0, void 0, function* () {
    return new Promise((resolve) => {
        let handshaked = false;
        const uri = `${host}?${qs.stringify(Object.assign({ verbose: 1 }, params))}`;
        const es = new eventsource_1.default(uri);
        const timeoutHandle = setTimeout(() => {
            es.close();
            console.error('connects to daemon timeout, please run "pipcook daemon restart".');
        }, 5000);
        const onerror = (e) => {
            if (handshaked === false) {
                es.close();
                clearTimeout(timeoutHandle);
                es.removeEventListener('error', onerror);
                // print error
                if (typeof e.status === 'number') {
                    console.error('occurrs an daemon error with the following response:\n', e);
                }
                else {
                    console.error(`daemon is not started(${e.message}), run "pipcook daemon start".`);
                }
            }
            else if (typeof handlers.error === 'function') {
                // manually pass the `error` event to user-defined handler.
                handlers.error(e);
            }
        };
        es.addEventListener('error', onerror);
        es.addEventListener('session', (e) => {
            if (e.data === 'close') {
                // close the connection and mark the handshaked is disabled.
                handshaked = false;
                es.close();
            }
            else if (e.data === 'start') {
                handshaked = true;
                // if `handlers.error` not defined, remove the listener directly.
                if (typeof handlers.error !== 'function') {
                    es.removeEventListener('error', onerror);
                }
                // clear the timeout handle because handshake is finished.
                clearTimeout(timeoutHandle);
                resolve(es);
            }
        });
        // register extra handlers.
        Object.keys(handlers)
            // handle `handlers.error` manually.
            .filter((name) => name !== 'error')
            .forEach((name) => {
            es.addEventListener(name, handlers[name]);
        });
    });
});
