// @pipcook/sdk/dist/index.js
"use strict";
var __createBinding = (this && this.__createBinding) || (Object.create ? (function(o, m, k, k2) {
    if (k2 === undefined) k2 = k;
    Object.defineProperty(o, k2, { enumerable: true, get: function() { return m[k]; } });
}) : (function(o, m, k, k2) {
    if (k2 === undefined) k2 = k;
    o[k2] = m[k];
}));
var __exportStar = (this && this.__exportStar) || function(m, exports) {
    for (var p in m) if (p !== "default" && !exports.hasOwnProperty(p)) __createBinding(exports, m, p);
};
Object.defineProperty(exports, "__esModule", { value: true });
exports.PipcookClient = void 0;
const job_1 = require("./job");
const pipeline_1 = require("./pipeline");
const plugin_1 = require("./plugin");
const request_1 = require("./request");
var utils_1 = require("./utils");
Object.defineProperty(exports, "JobStatusValue", { enumerable: true, get: function () { return utils_1.JobStatusValue; } });
Object.defineProperty(exports, "PluginStatusValue", { enumerable: true, get: function () { return utils_1.PluginStatusValue; } });
__exportStar(require("./interface"), exports);
/**
 * The Pipcook client to connect specific daemon endpoint.
 *
 * @example
 * ```js
 * const client = new PipcookClient();
 * const pipelines = await client.pipeline.list();
 * // [{ ...pipelines }]
 * ```
 */
class PipcookClient {
    /**
     * @param protocolWithHostname the daemon hostname with protocol, like "http://192.168.1.50"
     * @param port the port
     */
    constructor(protocolWithHostname = 'http://127.0.0.1', port = 6927) {
        this.endpoint = `${protocolWithHostname}:${port}/api`;
        this.pipeline = new pipeline_1.Pipeline(this.endpoint);
        this.job = new job_1.Job(this.endpoint);
        this.plugin = new plugin_1.Plugin(this.endpoint);
    }
    /**
     * list versions
     */
    listVersions() {
        return request_1.get(`${this.endpoint}/versions`);
    }
    /**
     * get daemon config
     */
    getConfig() {
        return request_1.get(`${this.endpoint}/config`);
    }
}
exports.PipcookClient = PipcookClient;
