// @pipcook/sdk/dist/plugin.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
exports.Plugin = void 0;
const request_1 = require("./request");
const base_1 = require("./base");
/**
 * API for plugin
 */
class Plugin extends base_1.BaseApi {
    constructor(url) {
        super(`${url}/plugin`);
    }
    /**
     * list all plugins
     * @param params params
     */
    list(params) {
        return request_1.get(`${this.route}`, params);
    }
    /**
     * get plugin by id
     * @param id string plugin id
     */
    get(id) {
        return request_1.get(`${this.route}/${id}`);
    }
    /**
     * get plugin metadata by id
     * @param id string plugin id
     */
    fetch(id) {
        return request_1.get(`${this.route}/${id}/metadata`);
    }
    /**
     * fetch specific plugin metadata by name
     * @param name the plugin name, for example "@pipcook/plugins-image-classification-data-collect"
     */
    fetchByName(name) {
        return request_1.get(`${this.route}/metadata`, { name });
    }
    /**
     * remove plugin or plugins
     * @param id string if null, remove all
     */
    remove(id) {
        return request_1.del(`${this.route}/${id ? id : ''}`);
    }
    /**
     * create by package name
     * @param name package name
     * @param pyIndex the python package index
     */
    createByName(name, pyIndex) {
        return request_1.post(`${this.route}`, { name, pyIndex });
    }
    /**
     * create by package stream
     * @param pkgStream file stream
     * @param pyIndex the python package index
     */
    createByTarball(pkgStream, pyIndex) {
        return request_1.uploadFile(`${this.route}/tarball`, pkgStream, { pyIndex });
    }
}
exports.Plugin = Plugin;
